pkg/healthz/healthz.go

```diff
@@ -0,0 +1,132 @@
+// Package healthz implements healthz handlers.
+package healthz
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"go.uber.org/zap"
+	pb "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1"
+	"sigs.k8s.io/aws-encryption-provider/pkg/plugin"
+)
+
+// ErrHandler is returned when the handler is configured wrong.
+var ErrHandler = errors.New("'encrypt' checker must precede 'decrypt' checker or 'encrypt' checker not found")
+
+// NewHandler returns a new healthz handler.
+func NewHandler(srv *plugin.Plugin, cli pb.KeyManagementServiceClient, healthzCheckers []string) (http.Handler, error) {
+	encryptCkFound := false
+	for _, ck := range healthzCheckers {
+		_, ok := checkers[ck]
+		if !ok {
+			return nil, fmt.Errorf("checker %q unknown", ck)
+		}
+		if ck == "encrypt" {
+			encryptCkFound = true
+			continue
+		}
+		if ck == "decrypt" && !encryptCkFound { // "encrypt" must precede
+			return nil, ErrHandler
+		}
+	}
+	return &handler{
+		mu:              sync.RWMutex{},
+		srv:             srv,
+		cli:             cli,
+		healthzCheckers: healthzCheckers,
+	}, nil
+}
+
+type handler struct {
+	mu              sync.RWMutex
+	srv             *plugin.Plugin
+	cli             pb.KeyManagementServiceClient
+	healthzCheckers []string
+	cipher          []byte
+}
+
+func (hd *handler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	res, err := plugin.Check(hd.cli)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(rw, http.StatusText(http.StatusInternalServerError))
+		zap.L().Error("failed healthz version", zap.Error(err))
+		return
+	}
+	zap.L().Info("passed healthz version", zap.String("version", res))
+	if len(hd.healthzCheckers) == 0 {
+		rw.WriteHeader(http.StatusOK)
+		fmt.Fprint(rw, http.StatusText(http.StatusOK))
+		return
+	}
+
+	for _, name := range hd.healthzCheckers {
+		zap.L().Info("checking", zap.String("name", name))
+		checker, ok := checkers[name]
+		if !ok {
+			rw.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprint(rw, http.StatusText(http.StatusInternalServerError))
+			zap.L().Warn("checker not found", zap.String("name", name))
+			return
+		}
+		err := checker(hd)
+		if err != nil {
+			rw.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprint(rw, fmt.Errorf("failed healthz KMS API (%v)", err))
+			return
+		}
+	}
+	rw.WriteHeader(http.StatusOK)
+	fmt.Fprint(rw, http.StatusText(http.StatusOK))
+	zap.L().Info("passed healthz KMS API")
+}
+
+var checkers = map[string]func(hd *handler) error{
+	"encrypt":      checkerEncrypt,
```

Comment:
I don't know that we need to expose all these knobs to users. I think for healthz, its either working or its not